### ISSUE TITLE
Does NCE loss in the code mismatch the Figure1?

### ISSUE THREAD

**Issue by hdjsjyl** [taesungp/contrastive-unpaired-translation#61]
In Figure1, I think the negatives are from patches of the same image/feature as the positive. But the code shows that the negatives are the combinations of vectors of positive. So I think the NCE loss code doesn't match the Figure1. Any advice would be appreciated, thanks!

**Comment by Shalomash** [comment]
I believe that the answer you are looking for is about where the negative/positive pairs are from. The positive pairs are the corresponding spatial 
location embeddings from the input and output. These spatial locations are created by reshaping the corresponding convolutional layer to (Batch_Size, H*W, number channels), the embeddings are then created by indexing a random spatial locations from this convolutional layer and passing it through a multilayer perceptron. The positive logits would be from the sum of cosine similarities from across the corresponding locations. This means the same locations are sampled to vectors that are then compared and the similarities are summed giving a (n_patches, 1) vector where the 1 axis has the value for the comparisons for the particular location. The negative pairs are taken from the same vectors that were used to compute the positive similarities. However, instead of comparing the corresponding locations, one of the matrix of the embeddings is transposed, yielding a (n_patches, n_patches) matrix. The identity of this matrix, however, would be comparisons between matching locations so this is set to -10.0 which goes to a 0 value when we are done computing the softmax. The logits are concatenated along the 1 axis and divided by the temperature and softmaxed yielding (n_patches, n_patches+1) where the 0 index for each of these comparisons is the positive logits and the target index.

**Comment by hdjsjyl** [comment]
Hi Shalomash, your answer totally addressed my issues. Thank you so much!
